ironfish/src/workerPool/tasks/decryptNotes.ts

```diff
@@ -120,42 +120,60 @@ export class DecryptNotesResponse extends WorkerMessage {
   }
 
   serializePayload(bw: bufio.StaticWriter | bufio.BufferWriter): void {
-    // TODO: the majority of responses will have 0 decrypted notes. It make sense to
-    // return a more compact serialization in that case.
-    for (const note of this.notes) {
-      const hasDecryptedNote = Number(!!note)
-      bw.writeU8(hasDecryptedNote)
-
-      if (note) {
-        let flags = 0
-        flags |= Number(!!note.index) << 0
-        flags |= Number(!!note.nullifier) << 1
-        flags |= Number(note.forSpender) << 2
-        bw.writeU8(flags)
-        bw.writeHash(note.hash)
-        bw.writeBytes(note.serializedNote)
-
-        if (note.index) {
-          bw.writeU32(note.index)
-        }
+    // The majority of responses will have 0 decrypted notes. A small
+    // percentage of responses will have a few decrypted notes. It's
+    // practically rare that a response will contain a large number of
+    // decrypted notes. For this reason, it makes sense to optimize for the
+    // case where the decrypted notes are 0 or close to 0.
+    //
+    // Here we use a sparse serialization: we write the length of the array at
+    // the beginning, and then we serialize only the non-null notes, prefixing
+    // them with their position in the array.
+    //
+    // In the most common case (0 decrypted notes), the total serialization
+    // size will be 4 bytes, irrespective of how many items the response has.
+    // In the second common case (few decrypted notes), the serialization size
+    // may still be smaller, or at least very close to, the size that a dense
+    // serialization would provide. In the most rare case (most/all decrypted
+    // notes), a sparse serialization has more overhead than a dense
+    // serialization, but that's an occurrence so rare and specific that we
+    // don't need to optimize for it.
+
+    bw.writeU32(this.notes.length)
+
+    for (const [arrayIndex, note] of this.notes.entries()) {
+      if (!note) {
+        continue
+      }
 
-        if (note.nullifier) {
-          bw.writeHash(note.nullifier)
-        }
+      bw.writeU32(arrayIndex)
+
+      let flags = 0
+      flags |= Number(!!note.index) << 0
+      flags |= Number(!!note.nullifier) << 1
+      flags |= Number(note.forSpender) << 2
+      bw.writeU8(flags)
+      bw.writeHash(note.hash)
+      bw.writeBytes(note.serializedNote)
+
+      if (note.index) {
+        bw.writeU32(note.index)
+      }
+
+      if (note.nullifier) {
+        bw.writeHash(note.nullifier)
       }
     }
   }
 
   static deserializePayload(jobId: number, buffer: Buffer): DecryptNotesResponse {
     const reader = bufio.read(buffer)
-    const notes = []
+
+    const arrayLength = reader.readU32()
+    const notes = Array(arrayLength).fill(null) as Array<DecryptedNote | null>
```

Comment:
do we need to return the `null` notes at all?

maybe i'm missing something but it looks like we always filter them out

Reply:
Yes: upper-level code needs to know which nodes exactly were successfully decrypted. Here's an example: suppose you want to decrypt all notes in this block:

```
block:
  transaction A: [note 1, note 2, note 3, note 4]
  transaction B: [note 5, note 6]
  transaction C: [note 7, note 8]
```

You can submit a decrypt request with all the notes:

```
[note 1, note 2, note 3, note 4, note 5, note 6, note 7, note 8]
```

Once you get a response, you need to know which notes were successfully decrypted, so that you can match them with the corresponding transactions.

If this response returned only non-null values, it wouldn't be possible to decrypt an entire block at once (or even multiple blocks).

Reply:
An alternative approach that I considered is to change the input from just `note` to `(block, transaction, note)`, and repeat the block/transaction information in the response. This however would greatly increase the size of requests, and that's why I didn't go that route.